lib/Producer.js

```diff
@@ -104,22 +104,35 @@ class Producer {
       options = {};
     }
 
-    if (!this.conns || this.conns.length === 0) {
-      return Promise.reject(new Error('No connections to nsqd')).asCallback(callback);
-    }
-
     if (options.retry && this.strategy === Producer.FAN_OUT) {
       return Promise.reject(new Error('Retry on produce level is not supported in fanout strategy')).asCallback(
         callback
       );
     }
 
+    if (!this.conns || this.conns.length === 0) {
+      // Start to reconnect to all the nsqd nodes if there is no available connection.
+      // Will follow the same retry strategy as specifed from the input parameter.
+      return Promise.resolve(
+        promiseRetry(options.retry === true ? {} : options.retry, (retry, number) => {
+          debug(`Attempt to reconnect to all nsqd nodes with retry number ${number}...`);
+          return this.connect().catch(retry);
+        }).then(() => {
+          return this._produce(topic, msg, options, callback);
+        })
+      ).asCallback(callback);
+    }
+
+    // When there are available nsqd connections.
+    return this._produce(topic, msg, options, callback);
+  }
+
+  _produce(topic, msg, options, callback) {
```

Comment:
what is the difference between `_produce` and `produce`? the params are exactly the same.
can't we just use `produce`?

Reply:
> what is the difference between `_produce` and `produce`? the params are exactly the same.
> can't we just use `produce`?

This is a good point. Originally I just want to refactor the non-connecting part of produce function out into _produce function, but actually it doesn't matter and we could continue call produce method once the connections are re-established. I will update the code according to this suggestion :)

Reply:
@xavierchow pr has been updated now to address this suggestion.